File: src/opochhash/proof/merkle.py
"""
Merkle Tree for Proof of Computation

Domain-separated, binary Merkle tree using OpochHash.
Supports:
- Building trees from leaf data
- Generating authentication paths (proofs)
- Verifying inclusion proofs
"""

from dataclasses import dataclass
from typing import List, Optional, Tuple
import hashlib

from .tags import ProofTag, tag_bytes


def _hash(tag: ProofTag, *parts: bytes) -> bytes:
    """Domain-separated hash using SHAKE256."""
    h = hashlib.shake_256()
    h.update(tag_bytes(tag))
    for part in parts:
        h.update(len(part).to_bytes(8, 'big'))
        h.update(part)
    return h.digest(32)


def leaf_hash(index: int, data: bytes) -> bytes:
    """Hash a leaf node: H(LEAF ‖ index ‖ data)"""
    return _hash(ProofTag.LEAF, index.to_bytes(8, 'big'), data)


def node_hash(left: bytes, right: bytes) -> bytes:
    """Hash an internal node: H(NODE ‖ left ‖ right)"""
    return _hash(ProofTag.NODE, left, right)


def root_hash(h: bytes) -> bytes:
    """Finalize a root: H(ROOT ‖ h)"""
    return _hash(ProofTag.ROOT, h)


@dataclass
class MerkleProof:
    """Authentication path for Merkle tree inclusion."""
    index: int
    leaf_data: bytes
    siblings: List[Tuple[bytes, bool]]  # (sibling_hash, is_right)

    def verify(self, root: bytes) -> bool:
        """Verify this proof against a root."""
        current = leaf_hash(self.index, self.leaf_data)

        for sibling, is_right in self.siblings:
            if is_right:
                current = node_hash(current, sibling)
            else:
                current = node_hash(sibling, current)

        return root_hash(current) == root

    def serialize(self) -> bytes:
        """Serialize proof to bytes."""
        parts = [
            self.index.to_bytes(8, 'big'),
            len(self.leaf_data).to_bytes(4, 'big'),
            self.leaf_data,
            len(self.siblings).to_bytes(4, 'big'),
        ]
        for sibling, is_right in self.siblings:
            parts.append(sibling)
            parts.append(bytes([1 if is_right else 0]))
        return b''.join(parts)

    @classmethod
    def deserialize(cls, data: bytes) -> 'MerkleProof':
        """Deserialize proof from bytes."""
        offset = 0

        index = int.from_bytes(data[offset:offset+8], 'big')
        offset += 8

        leaf_len = int.from_bytes(data[offset:offset+4], 'big')
        offset += 4
        leaf_data = data[offset:offset+leaf_len]
        offset += leaf_len

        num_siblings = int.from_bytes(data[offset:offset+4], 'big')
        offset += 4

        siblings = []
        for _ in range(num_siblings):
            sibling = data[offset:offset+32]
            offset += 32
            is_right = data[offset] == 1
            offset += 1
            siblings.append((sibling, is_right))

        return cls(index=index, leaf_data=leaf_data, siblings=siblings)


class MerkleTree:
    """
    Binary Merkle tree with domain-separated hashing.

    Properties:
    - Deterministic construction
    - O(log n) proof generation
    - O(log n) verification
    """

    def __init__(self, leaves: List[bytes]):
        """
        Build a Merkle tree from leaf data.

        Args:
            leaves: List of leaf data (will be hashed with indices)
        """
        if not leaves:
            raise ValueError("Cannot build empty Merkle tree")

        self.leaf_count = len(leaves)
        self.leaves = leaves

        # Hash leaves with their indices
        self.leaf_hashes = [
            leaf_hash(i, data) for i, data in enumerate(leaves)
        ]

        # Build tree bottom-up
        self.layers: List[List[bytes]] = [self.leaf_hashes]
        self._build_tree()

        # Finalize root
        self.root = root_hash(self.layers[-1][0])

    def _build_tree(self):
        """Build internal layers of the tree."""
        current_layer = self.leaf_hashes

        while len(current_layer) > 1:
            next_layer = []

            for i in range(0, len(current_layer), 2):
                left = current_layer[i]
                # If odd number of nodes, duplicate the last one
                right = current_layer[i + 1] if i + 1 < len(current_layer) else left
                next_layer.append(node_hash(left, right))

            self.layers.append(next_layer)
            current_layer = next_layer

    def get_proof(self, index: int) -> MerkleProof:
        """
        Generate authentication path for leaf at index.

        Args:
            index: Leaf index (0-based)

        Returns:
            MerkleProof that can verify inclusion
        """
        if index < 0 or index >= self.leaf_count:
            raise IndexError(f"Index {index} out of range [0, {self.leaf_count})")

        siblings = []
        current_index = index

        for layer in self.layers[:-1]:  # Exclude root layer
            # Determine sibling index
            if current_index % 2 == 0:
                # Current is left child, sibling is right
                sibling_index = current_index + 1
                is_right = True
            else:
                # Current is right child, sibling is left
                sibling_index = current_index - 1
                is_right = False

            # Handle case where sibling doesn't exist (odd layer)
            if sibling_index < len(layer):
                sibling_hash = layer[sibling_index]
            else:
                sibling_hash = layer[current_index]  # Duplicate

            siblings.append((sibling_hash, is_right))

            # Move to parent index
            current_index //= 2

        return MerkleProof(
            index=index,
            leaf_data=self.leaves[index],
            siblings=siblings
        )

    def verify_proof(self, proof: MerkleProof) -> bool:
        """Verify an inclusion proof against this tree's root."""
        return proof.verify(self.root)


def build_trace_tree(states: List[bytes]) -> MerkleTree:
    """
    Build a Merkle tree over an execution trace.

    Args:
        states: List of serialized states s_0, s_1, ..., s_T

    Returns:
        MerkleTree with root R committing to entire trace
    """
    return MerkleTree(states)


def verify_transition(
    root: bytes,
    proof_t: MerkleProof,
    proof_t1: MerkleProof,
    step_fn,
    program: bytes
) -> bool:
    """
    Verify a single transition in the trace.

    Args:
        root: Merkle root of trace
        proof_t: Proof for state at time t
        proof_t1: Proof for state at time t+1
        step_fn: Function (program, state) -> next_state
        program: Program bytecode

    Returns:
        True if transition is valid
    """
    # Verify both proofs
    if not proof_t.verify(root):
        return False
    if not proof_t1.verify(root):
        return False

    # Verify indices are consecutive
    if proof_t1.index != proof_t.index + 1:
        return False

    # Verify transition
    computed_next = step_fn(program, proof_t.leaf_data)
    return computed_next == proof_t1.leaf_data

File: opoch-poc-sha/rust-verifier/src/merkle.rs
//! Merkle Tree Implementation for STARK Commitments
//!
//! Uses SHA-256 with domain separation as specified.

use crate::sha256::Sha256;

/// Domain separation tags
const TAG_LEAF: u8 = 0x00;
const TAG_NODE: u8 = 0x01;
const TAG_ROOT: u8 = 0x02;

/// Compute leaf hash: SHA-256(0x00 || index_le64 || data)
pub fn leaf_hash(index: u64, data: &[u8]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(&[TAG_LEAF]);
    hasher.update(&index.to_le_bytes());
    hasher.update(data);
    hasher.finalize()
}

/// Compute node hash: SHA-256(0x01 || left || right)
pub fn node_hash(left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(&[TAG_NODE]);
    hasher.update(left);
    hasher.update(right);
    hasher.finalize()
}

/// Compute root hash: SHA-256(0x02 || h)
pub fn root_hash(h: &[u8; 32]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(&[TAG_ROOT]);
    hasher.update(h);
    hasher.finalize()
}

/// Merkle tree for commitments
pub struct MerkleTree {
    leaves: Vec<[u8; 32]>,
    layers: Vec<Vec<[u8; 32]>>,
    pub root: [u8; 32],
}

impl MerkleTree {
    /// Build Merkle tree from leaf data
    pub fn new(leaf_data: Vec<Vec<u8>>) -> Self {
        assert!(!leaf_data.is_empty(), "Cannot build empty tree");

        // Hash leaves with indices
        let leaves: Vec<[u8; 32]> = leaf_data
            .iter()
            .enumerate()
            .map(|(i, data)| leaf_hash(i as u64, data))
            .collect();

        // Pad to power of 2
        let n = leaves.len().next_power_of_two();
        let mut padded_leaves = leaves.clone();
        while padded_leaves.len() < n {
            padded_leaves.push([0u8; 32]);
        }

        // Build layers
        let mut layers = vec![padded_leaves.clone()];
        let mut current = padded_leaves;

        while current.len() > 1 {
            let mut next = Vec::with_capacity(current.len() / 2);
            for i in (0..current.len()).step_by(2) {
                let left = &current[i];
                let right = if i + 1 < current.len() {
                    &current[i + 1]
                } else {
                    left
                };
                next.push(node_hash(left, right));
            }
            layers.push(next.clone());
            current = next;
        }

        let root = root_hash(&current[0]);

        MerkleTree { leaves, layers, root }
    }

    /// Get authentication path for leaf at index
    pub fn get_path(&self, index: usize) -> MerklePath {
        let mut siblings = Vec::new();
        let mut current_index = index;

        for layer in &self.layers[..self.layers.len() - 1] {
            let sibling_index = if current_index % 2 == 0 {
                current_index + 1
            } else {
                current_index - 1
            };

            let sibling = if sibling_index < layer.len() {
                layer[sibling_index]
            } else {
                layer[current_index]
            };

            let is_right = current_index % 2 == 0;
            siblings.push((sibling, is_right));
            current_index /= 2;
        }

        MerklePath {
            index,
            siblings,
        }
    }

    /// Verify a path
    pub fn verify_path(&self, path: &MerklePath, leaf_data: &[u8]) -> bool {
        let mut current = leaf_hash(path.index as u64, leaf_data);

        for (sibling, is_right) in &path.siblings {
            if *is_right {
                current = node_hash(&current, sibling);
            } else {
                current = node_hash(sibling, &current);
            }
        }

        root_hash(&current) == self.root
    }
}

/// Merkle authentication path
#[derive(Clone, Debug)]
pub struct MerklePath {
    pub index: usize,
    pub siblings: Vec<([u8; 32], bool)>, // (sibling_hash, sibling_is_right)
}

impl MerklePath {
    /// Verify this path against a root
    pub fn verify(&self, leaf_data: &[u8], root: &[u8; 32]) -> bool {
        let mut current = leaf_hash(self.index as u64, leaf_data);

        for (sibling, is_right) in &self.siblings {
            if *is_right {
                current = node_hash(&current, sibling);
            } else {
                current = node_hash(sibling, &current);
            }
        }

        &root_hash(&current) == root
    }

    /// Serialize path
    pub fn serialize(&self) -> Vec<u8> {
        let mut result = Vec::new();
        result.extend_from_slice(&(self.index as u64).to_le_bytes());
        result.extend_from_slice(&(self.siblings.len() as u32).to_le_bytes());
        for (sibling, is_right) in &self.siblings {
            result.extend_from_slice(sibling);
            result.push(if *is_right { 1 } else { 0 });
        }
        result
    }

    /// Deserialize path
    pub fn deserialize(data: &[u8]) -> Option<Self> {
        if data.len() < 12 {
            return None;
        }

        let index = u64::from_le_bytes(data[0..8].try_into().ok()?) as usize;
        let num_siblings = u32::from_le_bytes(data[8..12].try_into().ok()?) as usize;

        let mut offset = 12;
        let mut siblings = Vec::with_capacity(num_siblings);

        for _ in 0..num_siblings {
            if offset + 33 > data.len() {
                return None;
            }
            let mut sibling = [0u8; 32];
            sibling.copy_from_slice(&data[offset..offset + 32]);
            let is_right = data[offset + 32] == 1;
            siblings.push((sibling, is_right));
            offset += 33;
        }

        Some(MerklePath { index, siblings })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_merkle_tree() {
        let leaves = vec![
            b"leaf0".to_vec(),
            b"leaf1".to_vec(),
            b"leaf2".to_vec(),
            b"leaf3".to_vec(),
        ];

        let tree = MerkleTree::new(leaves.clone());

        // Verify all paths
        for (i, leaf) in leaves.iter().enumerate() {
            let path = tree.get_path(i);
            assert!(tree.verify_path(&path, leaf));
            assert!(path.verify(leaf, &tree.root));
        }
    }

    #[test]
    fn test_path_serialization() {
        let leaves = vec![b"a".to_vec(), b"b".to_vec()];
        let tree = MerkleTree::new(leaves);
        let path = tree.get_path(0);

        let serialized = path.serialize();
        let deserialized = MerklePath::deserialize(&serialized).unwrap();

        assert_eq!(path.index, deserialized.index);
        assert_eq!(path.siblings.len(), deserialized.siblings.len());
    }
}
